### PR TITLE
ahoviewer: init at 1.4.6

### PR DESCRIPTION
**Pull request by skrzyp**
###### Things done:
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [x] Built on platform(s): NixOS / ~~OSX~~ / ~~Linux~~
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] ~~Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).~~ _(I'm not sure…)_
###### More

My first standalone Nixpkgs expression, please review it completely, because I'm unsure if I've done everything properly.

### CODE REVIEW

File: pkgs/applications/graphics/ahoviewer/default.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  buildInputs = [ pkgconfig glibmm libconfig gtkmm glibmm libxml2 gstreamer libsecret curl unrar libzip autoconf automake ];
+  preConfigure = ''
+    patchShebangs bootstrap
+    patchShebangs version.sh
+    ./bootstrap
+  '';
+  meta = {
+    homepage = "https://github.com/ahodesuka/ahoviewer";
+    description = "A GTK2 image viewer, manga reader, and booru browser";
+    license = "MIT";
```
Comment (hrdinka) [inline pkgs/applications/graphics/ahoviewer/default.nix]: Would you mind adding yourself as maintainer? The maintainers list can be found in `lib/maintainers.nix`.

File: pkgs/applications/graphics/ahoviewer/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+stdenv.mkDerivation rec {
+  name = "ahoviewer-1.4.6";
+  rev = "414cb91d66d96fab4b48593a7ef4d9ad461306aa";
+  src = fetchFromGitHub {
+    owner = "ahodesuka";
+    repo = "ahoviewer";
+    inherit rev;
+    sha256 = "081jgfmbwf2av0cn229cf4qyv6ha80ridymsgwq45124b78y2bmb";
+  };
+  buildInputs = [ pkgconfig glibmm libconfig gtkmm glibmm libxml2 gstreamer libsecret curl unrar libzip autoconf automake ];
```
Comment (hrdinka) [inline pkgs/applications/graphics/ahoviewer/default.nix]: If you want it to be 100% correct add `pkgconfig` `autoconf` `automake` to `nativeBuildInputs`. `buildInputs` are dependencies used by the package while `nativeBuildInputs` are needed to build the package. This is useful for cross builds (64bit host builds 32bit binaries) where `nativeBuildInputs` need to match the hosts architecture and `buildInputs` the one of the build target.

File: pkgs/applications/graphics/ahoviewer/default.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, pkgs, fetchurl, fetchFromGitHub, pkgconfig, libconfig, gtkmm, glibmm, libxml2, gstreamer, libsecret, curl, unrar, libzip, autoconf, automake }:
```
Comment (hrdinka) [inline pkgs/applications/graphics/ahoviewer/default.nix]: You might want to wrap long lines as these as well to make navigation through it easier. Most people here still wrap at 80 characters. But this is nothing crucial.

File: pkgs/applications/graphics/ahoviewer/default.nix [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+with stdenv.lib;
+stdenv.mkDerivation rec {
+  name = "ahoviewer-1.4.6";
+  rev = "414cb91d66d96fab4b48593a7ef4d9ad461306aa";
+  src = fetchFromGitHub {
+    owner = "ahodesuka";
+    repo = "ahoviewer";
+    inherit rev;
```
Comment (hrdinka) [inline pkgs/applications/graphics/ahoviewer/default.nix]: You aren't using `rev` anywhere outside this set. Therefore you can directly define it here. There is no need to define it directly in the package derivation if you don't need it anywhere else. But once again nothing crucial.

**Comment by skrzyp** [comment]
I've added myself as maintainer, but I have some issues with `configure.ac`, because it thinks that we don't have `gstreamer` (1.x is required, needed for WebM playback), but it's listed as `buildInputs`, can you help me with figuring out what's going there?

Also, `preConfigure` phase doesn't wok with `autoreconfHook` so I've used `prePatch`.

**Comment by hrdinka** [comment]
gstreamer 1.* is in gst_all_1, for example `gst_all_1.gstreamer` `gst_all_1.gstreamer`

**Comment by abbradar** [comment]
With this patch I got it working:
```
diff --git a/pkgs/applications/graphics/ahoviewer/default.nix b/pkgs/applications/graphics/ahoviewer/default.nix
index 1619603..a1fecd1 100644
--- a/pkgs/applications/graphics/ahoviewer/default.nix
+++ b/pkgs/applications/graphics/ahoviewer/default.nix
@@ -9,14 +9,18 @@ stdenv.mkDerivation {
     rev = "414cb91d66d96fab4b48593a7ef4d9ad461306aa";
     sha256 = "081jgfmbwf2av0cn229cf4qyv6ha80ridymsgwq45124b78y2bmb";
   };
-  nativeBuildInputs = [ autoreconfHook pkgconfig ];
+  nativeBuildInputs = [ autoreconfHook pkgconfig makeWrapper ];
   buildInputs = [ glibmm libconfig gtkmm glibmm libxml2 gst_all_1.gstreamer
-                  libsecret curl unrar libzip ];
-  propagatedBuildInputs = with gst_all_1; 
-                          [ gst-plugins-good gst-plugins-bad
-                            gst-libav ];
+                  libsecret curl unrar libzip
+                ]
+             ++ (with gst_all_1;
+                [ gst-plugins-base gst-plugins-bad ]);
   prePatch = ''patchShebangs version.sh'';
   configureFlags = "--enable-gst";
+  postInstall = ''
+    wrapProgram $out/bin/ahoviewer \
+      --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : "$GST_PLUGIN_SYSTEM_PATH_1_0"
+  '';
   meta = {
     homepage = "https://github.com/ahodesuka/ahoviewer";
     description = "A GTK2 image viewer, manga reader, and booru browser";
```
Haven't tested it extensively, though. Also I've noticed that `libunrar` was not found because we haven't built it -- I fixed that in https://github.com/NixOS/nixpkgs/commit/9e9a2f10966ab3889b523032ce83cf8c3b0698a8 .

**Comment by skrzyp** [comment]
`gst-plugins-{bad,good,libav}` are listed on project page as runtime dependencies, as opposite to `gstreamer-1` itself which is a compile-time dependency.

**Comment by skrzyp** [comment]
OK, using `gst-plugins-base` actually solved this problem, without doing any other crazy hacks around search paths.

This project was also reported to run on Windows, and its complexity is not too far to not be built on other system with GTK and GStreamer available, so only doubts are for OSX.

**Comment by abbradar** [comment]
Why do you use `propagatedBuildInputs`? It is needed if you want to propagate specified dependencies to packages which use _yours_ as a dependency -- it shouldn't have any effect on runtime dependencies.

If you need to add runtime GStreamer plugins, the proper way is to use `wrapProgram`. It may work without it but these plugins won't actually be available.

**Comment by skrzyp** [comment]
Because manual said so:

> If something is exclusively a build-time dependency, use buildInputs; if it’s (also) a runtime dependency, use propagatedBuildInputs.

**Comment by abbradar** [comment]
Ouch, the manual is misleading! We should definitely fix that, sorry for confusion.

What `propagatedBuildInputs` does is it adds listed dependencies to a file in `nix-support` directory. When later this package is listed in `buildInputs` somewhere, `stdenv` checks is there is anything in that file, and adds all packages listed there as implicit additional `buildInputs`.

On the other hand, there you need to specify to GStreamer how can it find its plugins -- usually it can be done via an environment variable, which you can "specify" for package via `makeWrapper` which sets it before running. Other example where `wrapProgram` can be useful is when your application tries to call some binary via `exec` -- so you need to add it to `PATH`.

**Comment by skrzyp** [comment]
It might work when I add `${gst_all_1.gst-plugins-base}` to `$PATH` via `makeWrapper` and ahoviewer itself can find them but can`t load them, when I tried to view some WebMs, any ideas?

**Comment by abbradar** [comment]
See my patch -- `gst-plugins-*` have setup hooks that populate `GST_PLUGIN_SYSTEM_PATH_1_0` at the build time, so if you set this variable in the wrapper like this:
```
  postInstall = ''
    wrapProgram $out/bin/ahoviewer \
      --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : "$GST_PLUGIN_SYSTEM_PATH_1_0"
  '';
```
it should work.

**Comment by skrzyp** [comment]
Will check it right now, got some free time, give me a sec.

**Comment by skrzyp** [comment]
Note about adding all of those `gst-plugins` at `buildInputs`, looks like Autotools can't detect gstreamer without them.

**Comment by abbradar** [comment]
That's the way it should be, yes -- but please move them to `buildInputs` and remove `propagatedBuildInputs` altogether -- it's not applicable here (again, sorry for the confusion from the manual!).

This doesn't install anything because you've replaced default `installPhase`  (which runs `make install`). The proper way to do this is via `postInstall` instead. Also, add `makeWrapper` to `nativeBuildInputs` (and to function arguments) -- without it `wrapProgram` won't be found.

**Comment by skrzyp** [comment]
Looks like it picked the wrong commit, sorry.

**Comment by abbradar** [comment]
LGTM, compiled and ran. One thing that you might want to fix before we merge this is this error on startup:
```
(ahoviewer:22496): glibmm-CRITICAL **: 
unhandled exception (type Glib::Error) in signal handler:
domain: gdk-pixbuf-error-quark
code  : 3
what  : Image type 'svg' is not supported
```
You can fix this by adding `librsvg` to build inputs and setting yet another environment variable:
```
diff --git a/pkgs/applications/graphics/ahoviewer/default.nix b/pkgs/applications/graphics/ahoviewer/default.nix
index ed00511..6ce757b 100644
--- a/pkgs/applications/graphics/ahoviewer/default.nix
+++ b/pkgs/applications/graphics/ahoviewer/default.nix
@@ -1,5 +1,5 @@
 { stdenv, pkgs, fetchurl, fetchFromGitHub, pkgconfig, libconfig, 
-  gtkmm, glibmm, libxml2, libsecret, curl, unrar, libzip, 
+  gtkmm, glibmm, libxml2, libsecret, curl, unrar, libzip, librsvg,
   gst_all_1, autoreconfHook, makeWrapper }:
 stdenv.mkDerivation {
   name = "ahoviewer-1.4.6";
@@ -12,7 +12,7 @@ stdenv.mkDerivation {
   enableParallelBuilding = true; 
   nativeBuildInputs = [ autoreconfHook pkgconfig makeWrapper ];
   buildInputs = [ glibmm libconfig gtkmm glibmm libxml2 
-                  libsecret curl unrar libzip 
+                  libsecret curl unrar libzip librsvg
                   gst_all_1.gstreamer
                   gst_all_1.gst-plugins-good 
                   gst_all_1.gst-plugins-bad 
@@ -21,7 +21,8 @@ stdenv.mkDerivation {
   postPatch = ''patchShebangs version.sh'';
   postInstall = ''
     wrapProgram $out/bin/ahoviewer \
-    --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : "$GST_PLUGIN_SYSTEM_PATH_1_0"
+    --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : "$GST_PLUGIN_SYSTEM_PATH_1_0" \
+    --set GDK_PIXBUF_MODULE_FILE "$GDK_PIXBUF_MODULE_FILE"
   '';
   meta = {
     homepage = "https://github.com/ahodesuka/ahoviewer";
```

**Comment by skrzyp** [comment]
What we'll do with `gnome-keyring` which is marked as runtime depenceny when we use `libsecret` features in this package? I think it's too intrusive as runtime dep, brings whole GTK3 stack and some background services to manage your keyrings (who even used it?), etc etc.

**Comment by abbradar** [comment]
I _think_ it should find Gnome Keyring via D-Bus so there's no need to add it to dependencies. Either way, we can defer this until someone has a problem (why would a manga reader/gallery browser need libsecret?) Let me test this one more time when I get home and I can push the button if you think it's ready. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by skrzyp** [comment]
> (why would a manga reader/gallery browser need libsecret?)

You might use this to store AP'I keys/passwords for some remote galleries.

**Comment by abbradar** [comment]
You haven't pulled my patch for `librsvg` -- what do we want to do with it?

**Comment by skrzyp** [comment]
Done, I'm wondering why my git repo lags one commit behind my local copy.

**Comment by abbradar** [comment]
Merged -- thanks for your work!

**Pull request merged**